privaterelay/tests/glean_tests.py

```diff
@@ -0,0 +1,568 @@
+    assert mask_data.has_website is False
+
+
+@pytest.mark.django_db
+def test_email_mask_relay_address_from_addon() -> None:
+    user = make_free_test_user()
+    mask = user.relayaddress_set.create(generated_for="example.com")
+
+    mask_data = EmailMaskData.from_mask(mask)
+
+    assert mask_data.has_website is True
+
+
+@pytest.mark.django_db
+def test_email_mask_domain_address() -> None:
+    user = make_premium_test_user()
+    user.profile.subdomain = "a_subdomain"
+    user.profile.save()
+    mask = user.domainaddress_set.create(address="custom")
+
+    mask_data = EmailMaskData.from_mask(mask)
+
+    assert mask_data.mask_id == mask.metrics_id
+    assert mask_data.is_random_mask is False
+    assert mask_data.has_website is False
+
+
+def create_expected_glean_payload(
+    category: str,
+    name: str,
+    extra_items: dict[str, str],
+    event_time: str,
+    app_channel: RELAY_CHANNEL_NAME,
+    telemetry_sdk_build: str,
+    ping_time: str,
+) -> dict[str, Any]:
+    """Return the expected payload, JSON-decoded from the glean log."""
+    return {
+        "metrics": {},
+        "events": [
+            create_expected_glean_event(category, name, extra_items, event_time)
+        ],
+        "client_info": {
+            "app_build": "Unknown",
+            "app_channel": app_channel,
+            "app_display_version": "2024.01.17",
+            "architecture": "Unknown",
+            "first_run_date": "Unknown",
+            "os": "Unknown",
+            "os_version": "Unknown",
+            "telemetry_sdk_build": telemetry_sdk_build,
+        },
+        "ping_info": {"seq": 0, "start_time": ping_time, "end_time": ping_time},
+    }
+
+
+def assert_glean_record(
+    record: LogRecord, user_agent: str = "", ip_address: str = ""
+) -> None:
+    """Check that the record is an expected glean record."""
+    assert record.name == "glean-server-event"  # Type in mozlog
+    assert record.msg == "glean-server-event"
+    assert record.levelname == "INFO"
+
+    # Check top-level extra data
+    assert getattr(record, "document_namespace") == "relay-backend"
+    assert getattr(record, "document_type") == "events"
+    assert getattr(record, "document_version") == "1"
+    assert UUID(getattr(record, "document_id")).version == 4
+    assert getattr(record, "user_agent") == user_agent
+    assert getattr(record, "ip_address") == ip_address
+    assert getattr(record, "payload").startswith("{")
+
+
+class PayloadVariedParts(NamedTuple):
+    """Parts of a Glean payload that vary over test runs"""
+
+    event_timestamp_ms: str
+    ping_time_iso: str
+    telemetry_sdk_build: str
+
+
+def extract_parts_from_payload(payload: dict[str, Any]) -> PayloadVariedParts:
+    """Check and return parts of the Glean payload that vary over test runs."""
+    event_ts_ms = payload["events"][0]["timestamp"]
+    event_time = datetime.fromtimestamp(event_ts_ms / 1000.0)
+    assert 0 < (datetime.now() - event_time).total_seconds() < 0.5
```

Comment:
question (non-blocking): why do we assert that the event time must have been less than 0.5 seconds ago?

Reply:
The timestamp should be in milliseconds. I'm trying to find a way to say "if we calculate the timestamp wrong, then it will be off by a factor of 1,000". This only tests the x1000 bigger, however...